---
title: "pcadapt"
output: html_notebook
---

Convert .vcf files (merged of all samples in each group) to .bed with plink

```{bash}
ml plink
plink --allow-extra-chr --double-id --vcf merged_deceased.vcf.gz --make-bed --out merged_deceased
plink --allow-extra-chr --double-id --vcf merged_survivors.vcf.gz --make-bed --out merged_survivors
```




```{r}
#install.packages("pcadapt")
library(pcadapt)
```

Reading in file, diagnostic plots, filtering outliers where p > 0.05, writing to a .csv for deceased birds.

```{r}

path_to_file <- "./merged_deceased.bed"


deceased_data<-read.pcadapt(path_to_file, type = "bed")
x <- pcadapt(input = deceased_data, K=4) 
x

#screeplot says first 2 PCs are best-or also 3 and 4? -retain pcs to left of straight line according to cattells rule
plot(x, option = "screeplot")

#making x have optimal number if PCS (2?)- rename to y
y<- pcadapt(input=deceased_data, K = 2)
summary(y)
y$pvalues


#score plot with first 2 PCs
plot(y, option = "scores")

#with PC 3 and 4
plot(x, i = 3, j = 4, option = "scores")


#manhattan plot

plot(y , option = "manhattan")

#qq plot- try with x and y

plot(y, option = "qqplot")

#x seems to look a bit better? 

#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

BiocManager::install("qvalue")

library(qvalue)
qval <- qvalue(y$pvalues)$qvalues
alpha <- 0.05
deceased_outliers <- which(qval < alpha)
length(deceased_outliers)
deceased_outliers<-outliers

#padj <- p.adjust(y$pvalues) # no correction needed
#alpha <- 0.05
#outliers <- which(padj < alpha)
#length(outliers)

#length(qval)

#print(NSoutliers)

deceased_outliers<-as.data.frame(deceased_outliers)
head(deceased_outliers)
#NSoutliers
write.csv(deceased_outliers,file="deceased_outliers_nocorrection.csv")
#outliers

#cat(NSoutliers)
```


Reading in file, diagnostic plots, filtering outliers where p > 0.05, writing to a .csv for surviving birds.

```{r}
library(pcadapt)

path_to_file <- "./merged_survivors.bed"


survivors_data<-read.pcadapt(path_to_file, type = "bed")
x <- pcadapt(input = survivors_data, K=4) 
#x

#screeplot says first 2 PCs are best-or also 3 and 4? -retain pcs to left of straight line according to cattells rule
plot(x, option = "screeplot")

#making x have optimal number if PCS (2?)- rename to y
y<- pcadapt(input=survivors_data, K = 2)
summary(y)
y$pvalues


#score plot with first 2 PCs
plot(y, option = "scores")

#with PC 3 and 4
plot(x, i = 3, j = 4, option = "scores")


#manhattan plot

plot(y , option = "manhattan")

#qq plot- try with x and y

plot(y, option = "qqplot")

#x seems to look a bit better? 

#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

BiocManager::install("qvalue")

library(qvalue)
qval <- qvalue(y$pvalues)$qvalues
alpha <- 0.05
survivors_outliers <- which(qval < alpha)
length(survivors_outliers)

survivors_outliers<-as.data.frame(survivors_outliers)
head(survivors_outliers)
write.csv(survivors_outliers,file="survivors_outliers_nocorrection.csv")

```


Linkage disequlibrium thinning of both groups.

```{r}
res<-pcadapt(survivors_data,K=4)
plot(res, option = "screeplot")

res <- pcadapt(survivors_data, K = 2)
plot(res)

par(mfrow = c(2, 2))
for (i in 1:2)
  plot(res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i)) #very clustered. LD might be an issue

#thinning survivors

res <- pcadapt(survivors_data, K = 10, LD.clumping = list(size = 200, thr = 0.1))
plot(res, option = "screeplot")

res <- pcadapt(sdata, K = 2, LD.clumping = list(size = 200, thr = 0.1))
par(mfrow = c(1, 2))
for (i in 1:2)
  plot(res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))

plot(res)

qval <- qvalue(res$pvalues)$qvalues
alpha <- 0.05
survivors_outliers_thinned <- which(qval < alpha)
length(survivors_outliers_thinned)

survivors_outliers_thinned<-as.data.frame(survivors_outliers_thinned)

write.csv(survivors_outliers_thinned,file="survivors_outliers_nocorrection_thinned.csv")
```

Deceased:

```{r}
non<-pcadapt(deceased_data,K=4)
plot(non, option = "screeplot")



par(mfrow = c(2, 2))
for (i in 1:2)
  plot(non$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i)) #very clustered. LD might be an issue

#thinning nonsurvivors

non <- pcadapt(deceased_data, K = 5, LD.clumping = list(size = 200, thr = 0.1))
plot(non, option = "screeplot")

non <- pcadapt(deceased_data, K = 2, LD.clumping = list(size = 200, thr = 0.1))
par(mfrow = c(1, 2))
for (i in 1:2)
  plot(non$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))

plot(non)

qval <- qvalue(non$pvalues)$qvalues
alpha <- 0.05
deceased_outliers_thinned <- which(qval < alpha)
length(deceased_outliers_thinned)

deceased_outliers_thinned<-as.data.frame(deceased_outliers_thinned)

write.csv(deceased_outliers_thinned, file="deceased_outliers_nocorrection_thinned.csv")
```

Matching outliers to list of variants in bash:

```{bash}

cut -d',' -f2 deceased_outliers_nocorrection.csv > d_outliers.txt #remove first column
cut -d',' -f2 survivors_outliers_nocorrection.csv > s_outliers.txt

sed -i '1d' d_outliers.txt #edit in place and remove header
sed -i '1d' s_outliers.txt

dos2unix d_outliers.txt #convert to unix line endings
dos2unix s_outliers.txt


awk 'NR==FNR{line[$1]; next} FNR in line' d_outliers.txt merged_deceased.bim > d_outliers_pcadapt.txt #match outliers to list of variants in bim file

awk 'NR==FNR{line[$1]; next} FNR in line' s_outliers.txt merged_survivors.bim > s_outliers_pcadapt.txt

```


Filtering the file to contain only the TLR positions of interest:

Read in:

```{r}
survivors <- read.table("s_outliers_pcadapt.txt")
deceased <- read.table("d_outliers_pcadapt.txt")
```

Positions:
tlr1a SUPER_4:53,881,681-53,884,137
tlr1b SUPER_4:53,873,395-53,875,356
TLR2A: SUPER_4:32,243,726-32,250,107
TLR2B: SUPER_4:32,234,631-32,240,982
TLR3: SUPER_4:20,097,425-20,107,088
tlr4 SUPER_18:8,037,993-8,042,694
tlr5 SUPER_3:95,457,095-95,459,677
TLR7: SUPER_1:77,565,800-77,578,017
TLR15: SUPER_3:116,325,458-116,332,055
TLR21: SUPER_29:1,577,922-1,588,503

For deceased:

```{r}
library(dplyr)
# rename columns to make life easier
#filter for chromosome and gene position, one chromosome at a time for simplicity

deceased_outliers_SUPER_4 <- deceased %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_4") %>%
  mutate(gene = case_when(between(pos, left=53881681, right=53884137) ~ "TLR1A", between(pos, left=53873395, right=53875356) ~ "TLR1B", between(pos, left=32243726, right=32250107) ~ "TLR2A", between(pos, left=32234631, right=32240982) ~ "TLR2B", between(pos, left=20097425, right=20107088) ~ "TLR3", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")


deceased_outliers_SUPER_18 <- deceased %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_18") %>%
  mutate(gene = case_when(between(pos, left=8037993, right=8042694) ~ "TLR4", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers


deceased_outliers_SUPER_3 <- deceased %>% #tlr5 SUPER_3:95,457,095-95,459,677 #TLR15: SUPER_3:116,325,458-116,332,055
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_3") %>%
  mutate(gene = case_when(between(pos, left=95457095, right=95459677) ~ "TLR5", between(pos, left=116325458, right=116332055) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers

#TLR7: SUPER_1:77,565,800-77,578,017
deceased_outliers_SUPER_1 <- deceased %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_7") %>%
  mutate(gene = case_when(between(pos, left=77565800, right=77578017) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

#TLR21: SUPER_29:1,577,922-1,588,503
deceased_outliers_SUPER_29 <- deceased %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_29") %>%
  mutate(gene = case_when(between(pos, left=1577922, right=1588503) ~ "TLR21", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

combined_deceased_outliers <- bind_rows(deceased_outliers_SUPER_1, deceased_outliers_SUPER_18, deceased_outliers_SUPER_29, deceased_outliers_SUPER_3, deceased_outliers_SUPER_4)

```


Repeat for survivors:

```{r}
survivors_outliers_SUPER_4 <- survivors %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_4") %>%
  mutate(gene = case_when(between(pos, left=53881681, right=53884137) ~ "TLR1A", between(pos, left=53873395, right=53875356) ~ "TLR1B", between(pos, left=32243726, right=32250107) ~ "TLR2A", between(pos, left=32234631, right=32240982) ~ "TLR2B", between(pos, left=20097425, right=20107088) ~ "TLR3", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")


survivors_outliers_SUPER_18 <- survivors %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_18") %>%
  mutate(gene = case_when(between(pos, left=8037993, right=8042694) ~ "TLR4", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers


survivors_outliers_SUPER_3 <- survivors %>% #tlr5 SUPER_3:95,457,095-95,459,677 #TLR15: SUPER_3:116,325,458-116,332,055
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_3") %>%
  mutate(gene = case_when(between(pos, left=95457095, right=95459677) ~ "TLR5", between(pos, left=116325458, right=116332055) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers

#TLR7: SUPER_1:77,565,800-77,578,017
survivors_outliers_SUPER_1 <- survivors %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_7") %>%
  mutate(gene = case_when(between(pos, left=77565800, right=77578017) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

#TLR21: SUPER_29:1,577,922-1,588,503
survivors_outliers_SUPER_29 <- survivors %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_29") %>%
  mutate(gene = case_when(between(pos, left=1577922, right=1588503) ~ "TLR21", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

combined_survivors_outliers <- bind_rows(survivors_outliers_SUPER_1, survivors_outliers_SUPER_18, survivors_outliers_SUPER_29, survivors_outliers_SUPER_3, survivors_outliers_SUPER_4)

# Add information about what group the sample is in 

combined_deceased_outliers <- combined_deceased_outliers %>%
  mutate(group = "deceased")

combined_survivors_outliers <- combined_survivors_outliers %>%
  mutate(group = "survivors")

#Combining both

pcadapt_outliers <- bind_rows(combined_deceased_outliers, combined_survivors_outliers)

#write to file
write.csv(pcadapt_outliers, file = "pcadapt_outliers.csv")

```



Repeat for LD thinned files:

```{r}

survivors_thinned <- read.table("s_outliers_pcadapt_thinned.txt")
deceased_thinned<- read.table("d_outliers_pcadapt_thinned.txt")

deceased_outliers_SUPER_4_thin <- deceased_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_4") %>%
  mutate(gene = case_when(between(pos, left=53881681, right=53884137) ~ "TLR1A", between(pos, left=53873395, right=53875356) ~ "TLR1B", between(pos, left=32243726, right=32250107) ~ "TLR2A", between(pos, left=32234631, right=32240982) ~ "TLR2B", between(pos, left=20097425, right=20107088) ~ "TLR3", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")


deceased_outliers_SUPER_18_thin <- deceased_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_18") %>%
  mutate(gene = case_when(between(pos, left=8037993, right=8042694) ~ "TLR4", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers


deceased_outliers_SUPER_3_thin <- deceased_thinned %>% #tlr5 SUPER_3:95,457,095-95,459,677 #TLR15: SUPER_3:116,325,458-116,332,055
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_3") %>%
  mutate(gene = case_when(between(pos, left=95457095, right=95459677) ~ "TLR5", between(pos, left=116325458, right=116332055) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers

#TLR7: SUPER_1:77,565,800-77,578,017
deceased_outliers_SUPER_1_thin <- deceased_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_7") %>%
  mutate(gene = case_when(between(pos, left=77565800, right=77578017) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

#TLR21: SUPER_29:1,577,922-1,588,503
deceased_outliers_SUPER_29_thin <- deceased_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_29") %>%
  mutate(gene = case_when(between(pos, left=1577922, right=1588503) ~ "TLR21", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

combined_deceased_outliers_thin <- bind_rows(deceased_outliers_SUPER_1_thin, deceased_outliers_SUPER_18_thin, deceased_outliers_SUPER_29_thin, deceased_outliers_SUPER_3_thin, deceased_outliers_SUPER_4_thin)

```



```{r}
survivors_outliers_SUPER_4_thin <- survivors_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_4") %>%
  mutate(gene = case_when(between(pos, left=53881681, right=53884137) ~ "TLR1A", between(pos, left=53873395, right=53875356) ~ "TLR1B", between(pos, left=32243726, right=32250107) ~ "TLR2A", between(pos, left=32234631, right=32240982) ~ "TLR2B", between(pos, left=20097425, right=20107088) ~ "TLR3", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")


survivors_outliers_SUPER_18_thin <- survivors_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_18") %>%
  mutate(gene = case_when(between(pos, left=8037993, right=8042694) ~ "TLR4", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers


survivors_outliers_SUPER_3_thin <- survivors_thinned %>% #tlr5 SUPER_3:95,457,095-95,459,677 #TLR15: SUPER_3:116,325,458-116,332,055
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_3") %>%
  mutate(gene = case_when(between(pos, left=95457095, right=95459677) ~ "TLR5", between(pos, left=116325458, right=116332055) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested") #no outliers

#TLR7: SUPER_1:77,565,800-77,578,017
survivors_outliers_SUPER_1_thin <- survivors_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_7") %>%
  mutate(gene = case_when(between(pos, left=77565800, right=77578017) ~ "TLR7", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

#TLR21: SUPER_29:1,577,922-1,588,503
survivors_outliers_SUPER_29_thin <- survivors_thinned %>%
  select(c(1,4,5,6)) %>%
  rename(chr=V1, pos=V4, base_before=V5, base_after=V6) %>%
  filter(chr=="SUPER_29") %>%
  mutate(gene = case_when(between(pos, left=1577922, right=1588503) ~ "TLR21", TRUE ~ "not_interested")) %>%
  filter(!gene=="not_interested")

combined_survivors_outliers_thinned <- bind_rows(survivors_outliers_SUPER_1_thin, survivors_outliers_SUPER_18_thin, survivors_outliers_SUPER_29_thin, survivors_outliers_SUPER_3_thin, survivors_outliers_SUPER_4_thin)

combined_deceased_outliers_thin <- combined_deceased_outliers_thin %>%
  mutate(group = "deceased")

combined_survivors_outliers_thinned <- combined_survivors_outliers_thinned %>%
  mutate(group = "survivors")

pcadapt_outliers_thinned <- bind_rows(combined_deceased_outliers_thin, combined_survivors_outliers_thinned)

write.csv(pcadapt_outliers_thinned, file = "pcadapt_outliers_thinned.csv")
```

